adapters/repos/db/vector/hnsw/sharded_node_locks.go

```diff
@@ -0,0 +1,66 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2023 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package hnsw
+
+import "sync"
+
+const nodeLockStripe = uint64(512)
+
+type shardedNodeLocks []*sync.RWMutex
+
+func newShardedNodeLocks() shardedNodeLocks {
+	snl := make([]*sync.RWMutex, nodeLockStripe)
+	for i := range snl {
+		snl[i] = new(sync.RWMutex)
+	}
+	return snl
+}
+
+func (snl shardedNodeLocks) LockAll() {
+	for i := range snl {
+		snl[i].Lock()
+	}
+}
+
+func (snl shardedNodeLocks) UnlockAll() {
+	for i := range snl {
+		snl[i].Unlock()
```

Comment:
Should we unlock them in reverse order to match the lock order? i.e. Can we end up in a deadlock if one of the lock is used somewhere else?

Reply:
good point. I will change that

Reply:
Mmm actually given how we are using them I can't think of any scenario where this could be a problem. Nevermind, it was more a doubt than an actual concern